Add more OWL tests

schivmeister opened the pull request:
All part of SC-R2, unless otherwise prefixed:

- Limit expressions to a select few properties
- PC-R3 Require `owl:versionInfo` on any instance of `owl:Ontology` matching SemVer pattern
- Minimum ontology header
- Advanced OWL axioms

The following UML-Derived rules are currently unimplementable due to tooling issues (tentatively reported [on Discord](https://discord.com/channels/911006583067144212/911020976496586782/1204449907226845194)):
1. CMC-R3 URI encoding characters
2. CMC-R4 URI naming scheme

schivmeister committed:
Add select few expressions test

We interpret the rule as recommending property expressions to be limited
to a select few, rather than recommending a minimum required set
(membership). However, the individual properties can be set to `minCount
1` to enable recommendations of those properties if such a minimum is
required.

This is the first use of a "closed" shape, which does not permit
properties other than those in the ignore list (which we currently
populate with the primary OWL object relations).

SEMIC rules: SC-R2

Co-authored-by: Ahmad, Jana, Ing <ahmadjan@fel.cvut.cz>

schivmeister committed:
Add semantic versioning test, sans versionIRI

We interpret the rule as alluding to (the required presence of)
`owl:versionInfo` when it prescribes a format and not (also)
`owl:versionIRI`. The IRI should ideally match the version in one form
or another (in any part of it) but that would be a more complex shape or
regular expression to validate. One place on the Web where both
properties are treated is at
https://obofoundry.org/principles/fp-004-versioning.html.

SEMIC rules: PC-R3

schivmeister committed:
Add minimum ontology header test

This is a derived subrule as the conventions do not explicitly ask for a
minimum recommended set of properties, but rather ask to be limited to a
few of them. We combine those with common usage for e.g. by ePO in order
to make some recommendations.

SEMIC rules: SC-R2 (partial)

P.S.: Language-tagged strings are different from plain strings and the
union value type cannot be checked. There is `dash:StringOrLangString`
but PySHACL appears to not have support for the DASH vocabulary. If we
were to make a DASH test case, however, it might've been validated with
`check_dash_result=True`.

Co-authored-by: Ahmad, Jana, Ing <ahmadjan@fel.cvut.cz>

schivmeister committed:
Add advanced definitions test

OWL restrictions are the key features of OWL2, which are typically
expressed through a blank node (bNode) typed as `a owl:Restriction`. For
these, we simply restrict usage of subsumption relationships to IRIs
and, for cases where named individuals are used and not bNodes, the
SC-R2 subrule/test `only_owl_resources` will raise a violation anyway
for types beyond the basics.

For the same reason (forbidden types), declaring property
characteristics (e.g. `a owl:TransitiveProperty`), or properties that
can only be used in a restriction (e.g. property constraints such as
`owl:someValuesFrom` and `owl:cardinality`) will raise the SC-R2
violation. For everything else that qualifies as "an advanced property",
be it whether used on a class or object property, we simply use a SHACL
alternative list and restrict cardinality to zero.

SEMIC rules: SC-R2 (partial)

schivmeister committed:
Add some exceptions to select_few_expressions

Forgotten `owl:imports` and `rfs:isDefinedBy` (used by ePO).

schivmeister committed:
Include header properties in select_few_expressions

There is overlap in the sense that there is a limited recommended set of
properties for an OWL ontology, and a minimum required set for an
ontology header.

schivmeister committed:
Restrict OWL ontology declaration to one instance

The shape did not include an upper limit as it couldn't be tested
together with the absence test case in a single file. However, it is
still good to have this restriction.

csnyulas reviewed (approved):
All looks fine, but consider any comments I added in the J column of the "4-01-2024 inventory for semic rules /  shacl shapes" spreadsheet.

schivmeister committed:
Change severity of select_few_expressions to Info

schivmeister committed:
Add upper cardinality to some of min_ontology_header

These make sense to have exact cardinalities but aren't tested as there
are other incorrect states in the same file (the addition of duplicate
values wouldn't raise the incorrect number and it is not important
enough to warrant a separate test data file).
